### DISCUSSION TITLE
Initial render in runtime using getStaticProps

### DISCUSSION THREAD

**Discussion by yanchumak** [vercel/next.js#40323] [Help]
Hi,

I am using  `getStaticProps` for SSG. My pages can't be built during build time(no internet access, private npm registry), it is possible only in runtime and **once** when the application starts.  `revalidate` mechanism is not the case. 

I've played with `notFound` flag and figured out that if `getStaticProps` returns `notFound: true` during build time, then in runtime `getStaticProps` will be called again where I can fetch the required data. After investigating the sources https://github.com/vercel/next.js/blob/40b0dae5586763e34c5d45ae5136ea1fa13c225d/packages/next/server/lib/incremental-cache/index.ts#L148 it has become clear why it works.

It looks like a bit tricky solution and I would like to know whether there is a proper way to do it.

**Comment by fitiskin** [comment]
Just an idea. Have you tried returning empty `paths` from `getStaticPaths` at build time (with `fallback: true`)? I guess this forces the rebuild for each new page at runtime (with `revalidate: false` by default).

**Reply by yanchumak** [reply]
yes, I have. it works fine, but I have many custom pages where each page has a specific structure and content. As far as I know, `getStaticPaths` is being used for one page where the difference is only resource id. For example post page `/posts/[id].js`.

**Reply by fitiskin** [reply]
Dynamic routes can be extended to catch all paths by adding three dots: `pages/post/[...slug].js`.

I haven't use it with `getStaticPaths` but seems like it should work:

> You can export `getStaticPaths` from a Dynamic Route that also uses `getStaticProps`.


Then you can resolve what to render or return `notFound` from `getStaticProps` depending on the `slug` value. It may not be very convenient but it should work.

**Reply by icyJoseph** [reply]
If your `getStaticPaths` function returns an empty list of paths, and fallback true, then at build time, only the fallback page will be generated.

**Reply by yanchumak** [reply]
Does it mean if I need something more complex, let's say `/posts/[postId]/comments/[commentId]` I need to parse URI and build my own routing logic?

**Reply by yanchumak** [reply]
what I want is just a flag that can disable during build time SSG in order to postpone this process in runtime - that's it.

**Comment by icyJoseph** [comment]
> My pages can't be built during build time(no internet access, private npm registry)

Does this mean that at build time, outgoing HTTP requests throw? Have you considered a little more resilient and arguably complete data fetching function?

```tsx
❌❌❌❌❌ I am afraid this doesn't really do the trick
import type { NextPage } from "next";
import Link from "next/link";

const Home: NextPage<{ ready: boolean; data: unknown; fatal: boolean }> = ({
  ready,
  data,
  fatal,
}) => {
  return (
    <div>
      Home <Link href="/blog">blog</Link>
      <div>more stuff</div>
    </div>
  );
};

export const getStaticProps = async () => {
  try {
    const res = await fetch("https://unreachable-at-build-time");
    if (res.status === 404) return { notFound: true, revalidate: 1000 };
    const data = await res.json();
    return { props: { ready: true, data }, revalidate: 100 };
  } catch (e) {
    if (e instanceof TypeError) {
      return { props: { ready: false }, revalidate: 100 };
    }
    return { props: { fatal: true }, revalidate: 100 };
  }
};

export default Home;
```

Then if the page is not ready, you can show a loading spinner of some sort... For fatal errors perhaps show something else, and otherwise, show data or a 404 page.

**Reply by yanchumak** [reply]
If during build time there is no internet we go to catch, in runtime getStaticProps won't be called and we have loading spinner forever.

**Reply by icyJoseph** [reply]
Woops, forgot the revalidate property... my bad, fixed example - Still I think it won't cut it right? It'll just hang on loading for a few users mmmm and then the client has to somehow trigger a page load... ugh, not so nice

**Comment by icyJoseph** [comment]
Alright, another hack would be to scope this pages to a sub directory `/pages/[runtime]/hello` for example, and then these pages would look like:

```tsx
import type { GetStaticPaths, GetStaticProps, NextPage } from "next";

import { useRouter } from "next/router";

export const FooPage: NextPage<{
  foo: string;
}> = (props) => {
  const { isFallback } = useRouter();

  if (isFallback) return <div>Loading...</div>;

  return <div>{props.foo}</div>;
};

export const getStaticPaths: GetStaticPaths<{ page: string }> = () => {
  return {
    paths: [],
    fallback: true,
  };
};

export const getStaticProps: GetStaticProps = async (ctx) => {
  const runtime = ctx.params?.runtime;

  if (typeof runtime === "undefined") return { notFound: true };
  if (Array.isArray(runtime)) return { notFound: true };

  return {
    props: {
      foo: "bar",
    },
  };
};

export default FooPage;
```

This page won't be build at build time, because of the empty paths list. 

Then in `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  async rewrites() {
    return {
      beforeFiles: [{ source: "/hello", destination: "/runtime/hello" }],
    };
  },
};

module.exports = nextConfig;

```

If you want to prevent direct access to `/runtime/hello`, then it is better to use `middleware` to conduct rewrites.

> runtime in this case is just trivial, the idea is just to force getStaticPaths to get involved

**Reply by yanchumak** [reply]
Interesting example, looks more complex than just returning `notFound` during build time=)
Does my request very specific and has no one faced before with such a case?

**Reply by icyJoseph** [reply]
The thing is that `getServerSideProps` is slightly better for that use case, https://nextjs.org/docs/going-to-production#caching, but it requires you to set cache control headers, and possibly add a caching layer:

> Your deployment provider must support caching for dynamic responses. If you are self-hosting, you will need to add this logic yourself using a key/value store like Redis. If you are using Vercel, [Edge Caching works without configuration](https://vercel.com/docs/edge-network/caching?utm_source=next-site&utm_medium=docs&utm_campaign=next-website).

All in all, we'd want to avoid hitting the actual `GSSP` function as much as possible, by telling the the CDN to revalidate over long periods of time.
